astropy/utils/decorators.py

```diff
@@ -273,6 +275,241 @@ def delete(self):
+        # optional parameters may not be set, so if these are not iterables
+        # wrap them.
+        if not isinstance(arg_in_kwargs, cls_iter):
+            arg_in_kwargs = [arg_in_kwargs] * n
+        if not isinstance(relax, cls_iter):
+            relax = [relax] * n
+    else:
+        # To allow a uniform approach later on, wrap all arguments in lists.
+        n = 1
+        old_name = [old_name]
+        new_name = [new_name]
+        since = [since]
+        arg_in_kwargs = [arg_in_kwargs]
+        relax = [relax]
+
+    def decorator(function):
+        # Lazy import to avoid cyclic imports
+        from .compat.funcsigs import signature
+
+        # The named arguments of the function.
+        arguments = signature(function).parameters
+        keys = list(arguments.keys())
+        position = [None] * n
+
+        for i in range(n):
+            # Determine the position of the argument.
+            if new_name[i] in arguments:
+                param = arguments[new_name[i]]
+                # There are several possibilities now:
+
+                # 1.) Positional or keyword argument:
+                if param.kind == param.POSITIONAL_OR_KEYWORD:
+                    position[i] = keys.index(new_name[i])
+
+                # 2.) Keyword only argument (Python 3 only):
+                elif param.kind == param.KEYWORD_ONLY:
+                    # These cannot be specified by position.
+                    position[i] = None
+
+                # 3.) positional-only argument, varargs, varkwargs or some
+                #     unknown type:
+                else:
+                    raise TypeError('cannot replace argument "{0}" of kind {1}'
+                                    '.'.format(new_name[i], repr(param.kind)))
+
+            # In case the argument is not found in the list of arguments
+            # the only remaining possibility is that it should be catched
+            # by some kind of **kwargs argument.
+            # This case has to be explicitly specified, otherwise throw
+            # an exception!
+            elif arg_in_kwargs[i]:
+                position[i] = None
+            else:
+                raise TypeError('"{}" was not specified in the function '
+                                'signature. If it was meant to be part of '
+                                '"**kwargs" then set "arg_in_kwargs" to "True"'
+                                '.'.format(new_name[i]))
+
+        if function.__doc__:
+            # Remove indentation of the original documentation so it will be
+            # correctly rendered.
+            doc = [textwrap.dedent(function.__doc__).strip('\n')]
+
+            # Append the deprecation messages ordered by their "since".
+            def item2(x):
+                return LooseVersion(str(x[2]))
+
+            sorted_zipped = sorted(zip(old_name, new_name, since), key=item2)
+            for k, vals in itertools.groupby(sorted_zipped, key=item2):
```

Comment:
Just for the ease of future maintenance by someone else, perhaps explain why there is a need to sort first and then do a grouping, both using the same keys.